### PR TITLE
update OverrideConfig type and add listers to localizer

### PR DESCRIPTION
**Pull request by qmhu**
#### What type of PR is this?
Overrider

#### What this PR does / why we need it:
Implement a Overrider to support jsonpatch and smp for resources

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: pkg/hub/localizer/helm.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package localizer
```
Comment (dixudx) [inline pkg/hub/localizer/helm.go:1]: add license

File: pkg/hub/localizer/helm.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"encoding/json"
+
+	"github.com/ghodss/yaml"
+	"helm.sh/helm/v3/pkg/chartutil"
+)
+
+// MergeChartValues merges src and override values into merged values for helm chart
+func MergeChartValues(overrides, src []byte) ([]byte, error) {
```
Comment (dixudx) [inline pkg/hub/localizer/helm.go:11]: add tests

File: pkg/hub/localizer/jsonpatch.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package localizer
```
Comment (dixudx) [inline pkg/hub/localizer/jsonpatch.go:1]: add license

File: pkg/hub/localizer/jsonpatch.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/ghodss/yaml"
+)
+
+type JsonPatcher struct {
+	Raw         []byte
+	PatchObject jsonpatch.Patch
+}
+
+func (p *JsonPatcher) Apply() ([]byte, error) {
```
Comment (dixudx) [inline pkg/hub/localizer/jsonpatch.go:15]: add tests

File: pkg/hub/localizer/localizer_test.go [hunk 5]
```diff
@@ -0,0 +1,412 @@
+package localizer
+
+import (
+	"github.com/ghodss/yaml"
```
Comment (dixudx) [inline pkg/hub/localizer/localizer_test.go:4]: format import

File: pkg/hub/localizer/localizer_test.go [hunk 6]
```diff
@@ -0,0 +1,412 @@
+package localizer
```
Comment (dixudx) [inline pkg/hub/localizer/localizer_test.go:1]: add license

File: pkg/hub/localizer/localizer_test.go [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+import (
+	"github.com/ghodss/yaml"
+	"reflect"
+	"testing"
+
+	appsapi "github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+)
+
+func TestPatchWithSmp(t *testing.T) {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer_test.go:11]: move this test to smp_test.go

File: pkg/hub/localizer/smp.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
+package localizer
```
Comment (dixudx) [inline pkg/hub/localizer/smp.go:1]: add license

File: pkg/hub/localizer/smp.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"encoding/json"
```
Comment (dixudx) [inline pkg/hub/localizer/smp.go:4]: add an empty line here

File: pkg/hub/localizer/localizer.go [hunk 10]
```diff
@@ -99,7 +112,282 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+func (l *Localizer) getGlobalizationByResource(unstructured *unstructured.Unstructured) ([]*appsapi.Globalization, error) {
+	var glos []*appsapi.Globalization
+
+	globalizations, err := l.globLister.List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, glo := range globalizations {
+		if IsFeedMatchObject(glo.Spec.Feed, unstructured) {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:259]: use functions in pkg/utils/feed.go

File: pkg/hub/localizer/localizer.go [hunk 11]
```diff
@@ -99,7 +112,282 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+
+			if len(localizations) <= 0 {
+				klog.V(4).Info("cannot found Localizations matching for Description %s", klog.KObj(desc))
+			} else {
+				for _, loc := range localizations {
+					outputObject, err = l.executionOverride(appsapi.GenericOverride, loc.Spec.Overrides, outputObject)
+					if err != nil {
+						msg := fmt.Sprintf("failed to override for Description %s and Localization %s: %v", klog.KObj(desc), klog.KObj(loc), err)
+						klog.Error(msg)
+						l.recorder.Event(desc, corev1.EventTypeWarning, "FailedLocalizationObject", msg)
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:236]: s/FailedLocalizationObject/FailedOverridingObject

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/localizer/localizer.go (hunk 10 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer.go:259]: use the label selector to implement the getGlobalizationByxxx function

File: pkg/hub/localizer/helm.go (hunk 1 above)
Comment (dixudx) [reply pkg/hub/localizer/helm.go:17]: still outstanding.

File: pkg/hub/localizer/localizer.go [hunk 12]
```diff
@@ -20,7 +20,12 @@ import (
 	"context"
 	"fmt"
 
+	"encoding/json"
+	"errors"
+	corev1 "k8s.io/api/core/v1"
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:26]: Please format your imports. Thanks.

File: pkg/hub/localizer/localizer.go [hunk 13]
```diff
@@ -32,6 +37,12 @@ import (
 	applisters "github.com/clusternet/clusternet/pkg/generated/listers/apps/v1alpha1"
 	"github.com/clusternet/clusternet/pkg/known"
 	"github.com/clusternet/clusternet/pkg/utils"
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/ghodss/yaml"
+)
+
+var (
+	ErrMissingValue = errors.New("missing value")
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:47]: This var is used only once. No need to declare it.

**Review by dixudx — CHANGES_REQUESTED** [review]
Please rebase with main brach. Thanks.

File: pkg/hub/localizer/localizer.go (hunk 12 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer.go:26]: cannot reproduce, will recheck in newest commit

File: pkg/hub/localizer/localizer.go (hunk 13 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer.go:47]: dome

File: pkg/hub/localizer/localizer_test.go [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+
+		toYaml, _ := yaml.JSONToYAML(output)
+		if string(toYaml) != string(tt.want) {
+			t.Errorf("%v: Got %s WANT %s", tt.testName, string(output), string(tt.want))
+		}
+	}
+}
+
+func TestMergeChartValues(t *testing.T) {
+	src := `{"replicas":"3","image":"nginx:v0.6.0","a":{"b":"c"}}`
```
Comment (dixudx) [inline pkg/hub/localizer/localizer_test.go]: better to keep `src`, `input`, `want` with the same syntax, `yaml` or `json`.

File: pkg/hub/localizer/localizer_test.go [hunk 15]
```diff
@@ -0,0 +1,412 @@
+package localizer
+
+import (
+	"github.com/ghodss/yaml"
+	"reflect"
```
Comment (dixudx) [inline pkg/hub/localizer/localizer_test.go]: Please format the imports. Still outstanding.
Please don't make this resolved unless it is addressed.

File: pkg/hub/localizer/localizer.go [hunk 16]
```diff
@@ -122,7 +133,282 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+	}
+
+	return locs, nil
+}
+
+func (l *Localizer) executionOverride(overriderType appsapi.OverrideType, overrides []appsapi.OverrideRule, raw []byte) ([]byte, error) {
+	output := raw
+	var err error
+	for _, override := range overrides {
+		if overriderType == appsapi.GenericOverride {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go]: Could it be better we could aggregate all the overrides, and then apply final override once.

File: pkg/hub/localizer/localizer.go [hunk 17]
```diff
@@ -122,7 +133,282 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+func (l *Localizer) getLocalizationByResource(namespace string, unstructured *unstructured.Unstructured) ([]*appsapi.Localization, error) {
+	var locs []*appsapi.Localization
+
+	localizations, err := l.locLister.Localizations(namespace).List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, loc := range localizations {
+		if IsFeedMatchObject(loc.Spec.Feed, unstructured) {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go]: Same

File: pkg/hub/localizer/localizer.go [hunk 18]
```diff
@@ -122,7 +133,282 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+			return false
+		}
+
+		return feedSelector.Matches(labels.Set(unstructured.GetLabels()))
+	}
+
+	return false
+}
+
+func IsFeedMatchHelmChart(feed appsapi.Feed, chartReference appsapi.ChartReference) bool {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go]: what if we specify `feed.Selector` here?

File: pkg/hub/localizer/localizer_test.go (hunk 15 above)
Comment (dixudx) [reply pkg/hub/localizer/localizer_test.go]: Again.
Please add license here.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/localizer/localizer.go (hunk 18 above)
Comment (dixudx) [reply pkg/hub/localizer/localizer.go]: For every `Localization` and `Globalization`, you could find a label, like `some-uid-string=Namespace`.

File: pkg/hub/localizer/localizer.go [hunk 19]
```diff
@@ -122,7 +142,257 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+			string(manifest[0].UID): unstructured.GetKind(),
+		}))
+		if err != nil {
+			msg := fmt.Sprintf("failed to get Globalizations matching for Description %s: %v", klog.KObj(desc), err)
+			klog.Error(msg)
+			l.recorder.Event(desc, corev1.EventTypeWarning, "FailedRetrievingGlobalization", msg)
+			return err
+		}
+
+		if len(globalizations) <= 0 {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:272]: could not be less than 0

File: pkg/hub/localizer/localizer.go [hunk 20]
```diff
@@ -122,7 +142,257 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+			string(chart.GetUID()): chart.Kind,
+		}))
+		if err != nil {
+			msg := fmt.Sprintf("failed to get Globalizations matching for Description %s: %v", klog.KObj(desc), err)
+			klog.Error(msg)
+			l.recorder.Event(desc, corev1.EventTypeWarning, "FailedRetrievingGlobalization", msg)
+			return err
+		}
+
+		if len(globalizations) <= 0 {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:177]: Here `globalizations` could only be `nil` or non-nil ( an array)

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/localizer/helm.go [hunk 21]
```diff
@@ -0,0 +1,37 @@
+package localizer
+
```
[the remark this answers is not in the document]
Comment (qmhu) [reply pkg/hub/localizer/helm.go:18]: not exist in newest version

File: pkg/hub/localizer/localizer_test.go (hunk 14 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer_test.go]: ok, I will normalize all cases except one exception:  MergeChartValues can accept yaml style input ,

File: pkg/hub/localizer/localizer.go (hunk 18 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer.go]: maybe it's the old version problem, using the label selector solution now

File: pkg/hub/localizer/localizer.go (hunk 16 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer.go]: Customer may use different patch strategy in the same Localization. for each override item, we need to parser the values then know the patch part is a JsonPatch or a SMP Patch, so execute override one by one should be easier here.
`apiVersion: apps.clusternet.io/v1alpha1
kind: Localization
metadata:
  name: foo-loc
  namespace: clusternet-px5n9
spec:
  overrides:
    - ruleName: addLabel2
      overrideValue:
        patch: |-
          apiVersion: v1
          kind: Namespace
          metadata:
            name: foo
            labels:
              clusternet.io/localization-injected: true
    - ruleName: addAnno
      overrideValue:
        patch: |-
          - path: "/metadata/annotations"
            op: "add"
            value:
              app: apple
  feed:
    apiVersion: v1
    kind: Namespace
    name: foo`

Comment (qmhu) [reply pkg/hub/localizer/localizer.go]: It's more clear to aggregate HelmChart values first and merge values finally.

File: pkg/hub/localizer/localizer.go [hunk 22]
```diff
@@ -29,8 +36,11 @@ import (
 	"github.com/clusternet/clusternet/pkg/controllers/apps/localization"
 	clusternetclientset "github.com/clusternet/clusternet/pkg/generated/clientset/versioned"
 	clusternetinformers "github.com/clusternet/clusternet/pkg/generated/informers/externalversions"
+	applisters "github.com/clusternet/clusternet/pkg/generated/listers/apps/v1alpha1"
 	"github.com/clusternet/clusternet/pkg/known"
 	"github.com/clusternet/clusternet/pkg/utils"
+	jsonpatch "github.com/evanphx/json-patch"
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:42]: Move these two upper

File: pkg/hub/localizer/localizer.go [hunk 23]
```diff
@@ -39,6 +49,11 @@ type Localizer struct {
 
 	clusternetClient *clusternetclientset.Clientset
 
+	locLister      applisters.LocalizationLister
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:52]: `locSynced cache.InformerSynced`

And same to below three listers.

File: pkg/hub/localizer/helm.go [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+	"encoding/json"
+
+	"github.com/ghodss/yaml"
+	"helm.sh/helm/v3/pkg/chartutil"
+)
+
+// MergeChartValues merges src and override values into merged values for helm chart
+func MergeChartValues(overrides, src []byte) ([]byte, error) {
+	var srcValues map[string]interface{}
+	if src != nil {
```
Comment (dixudx) [inline pkg/hub/localizer/helm.go:29]: no need to check src is nil or not, `json.Unmarshal` will do it.

File: pkg/hub/localizer/helm.go [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+		err := json.Unmarshal(src, &srcValues)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var overrideValues map[string]interface{}
+	if overrides != nil {
```
Comment (dixudx) [inline pkg/hub/localizer/helm.go:37]: Same

File: pkg/hub/localizer/localizer.go (hunk 22 above)
Comment (qmhu) [reply pkg/hub/localizer/localizer.go:42]: ok, I misunderstood the import rules...

File: pkg/hub/localizer/helm.go [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+package localizer
+
+import (
+	"encoding/json"
+
+	"github.com/ghodss/yaml"
+	"helm.sh/helm/v3/pkg/chartutil"
+)
+
+// MergeChartValues merges src and override values into merged values for helm chart
```
Comment (dixudx) [inline pkg/hub/localizer/helm.go:26]: Could the input params just be `map[string]interface`?

For caller `aggregatedValues, err = MergeChartValues([]byte(override.OverrideValue["values"]), aggregatedValues)`, no need to have extra byte operations.

File: pkg/hub/localizer/helm_test.go [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+import "testing"
+
+func TestMergeChartValues(t *testing.T) {
+	tests := []struct {
+		name  string
+		src   []byte
+		input []byte
+		want  []byte
+	}{
+		{
```
Comment (dixudx) [inline pkg/hub/localizer/helm_test.go:28]: add a complex testcase for nested override.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/hub/localizer/localizer.go [hunk 28]
```diff
@@ -122,7 +142,258 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+	}
+
+	patchObj, err := jsonpatch.DecodePatch(patchJSON)
+	if err == nil {
+		return &JsonPatcher{
+			PatchObject: patchObj,
+			Raw:         raw,
+		}, nil
+	} else {
+		return &SmpPatcher{
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:388]: This is not a good way to handle this.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/apis/apps/v1alpha1/localization.go [hunk 29]
```diff
@@ -75,17 +75,38 @@ const (
...
+// Available fields for OverrideType are: JSONPatch, StrategicMergePatch and HelmValues.
+const (
+	// JsonPatchOverride apply JSONPatch for all matched objects (excluding HelmChart).
+	JSONPatchOverride OverrideType = "JSONPatch"
+
+	// StrategicMergePatchOverride apply Strategic MergePatch for all matched objects (excluding HelmChart).
+	StrategicMergePatchOverride OverrideType = "StrategicMergePatch"
+
+	// HelmValuesOverride apply HelmValues for all matched HelmCharts.
+	HelmValuesOverride OverrideType = "HelmValues"
```
Comment (dixudx) [inline pkg/apis/apps/v1alpha1/localization.go:89]: s/HelmValues/Helm

File: pkg/hub/localizer/helm.go [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+func MergeChartValues(overrides []byte, src map[string]interface{}) (map[string]interface{}, error) {
+	var overrideValues map[string]interface{}
+	err := yaml.Unmarshal(overrides, &overrideValues)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal patch %s: %v", overrides, err)
+	}
+
+	mergedValues := chartutil.CoalesceTables(make(map[string]interface{}, len(overrideValues)), overrideValues)
```
Comment (dixudx) [inline pkg/hub/localizer/helm.go:34]: `mergedValues := chartutil.CoalesceTables(map[string]interface{}{}, overrideValues)`

File: pkg/hub/localizer/helm_test.go [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+app: bar
+replicas: 3
+nested:
+  app: apple
+top:
+  sub:
+    sword: wood
+    foo: bar
+`)
+	output, err := MergeChartValues(input, src)
```
Comment (dixudx) [inline pkg/hub/localizer/helm_test.go:95]: compare `output` with `want` ?

File: pkg/hub/localizer/localizer.go [hunk 32]
```diff
@@ -122,7 +168,205 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
...
+	}
+
+	desc.Spec.Raw = outputObjects
 	return nil
 }
+
+func (l *Localizer) getAllOverridesFromDescription(uid string, kind string, desc *appsapi.Description) ([]appsapi.OverrideConfig, error) {
+	var allOverrides []appsapi.OverrideConfig
+
+	cluster, err := l.clusterLister.ManagedClusters(desc.Namespace).Get(desc.Labels[known.ClusterNameLabel])
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:261]: I don't think `clusterLister` is needed.

File: pkg/hub/localizer/smp.go [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+
+var lock sync.Mutex
+
+// SmpPatcher implements StrategicMergePatch for apply a Patch object to a Kubernetes resource
+type SmpPatcher struct {
+	Src   []byte //Kubernetes resource with JSON format
+	Patch []byte //Kubernetes resource with JSON or YAML format
+}
+
+func (p *SmpPatcher) Apply() ([]byte, error) {
```
Comment (dixudx) [inline pkg/hub/localizer/smp.go:36]: weird implementation.

**Pull request merged**